### PR TITLE
[Feature] Side-by-side view for plagiarism inspection (Modeling Exercises)

### PR DESCRIPTION
**Pull request by philippbauch**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I added screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
For motivation and context, check out #2100.

### Description
I added the `ArtemisPlagiarismModule` under `exercises/shared` which exports the `PlagiarismInspectorComponent`.

To implement the `PlagiarismSplitViewComponent`, I added the [split.js](https://split.js.org/) library.

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration > Your Course > Exercises
3. Create a modeling exercise
4. Use multiple accounts to create similar submissions for your modeling exercise
5. Go to the exercise details page
6. Click 'Check plagiarism'

You should now be redirected to a new page with the plagiarism inspector. Check the following functionality:

1. Given similar submissions exist, clicking on 'Check Plagiarism' should fetch detected plagiarisms and display them in the sidebar on the left
2. Detected plagiarisms are sorted by their similarity in descending order
2. Selecting a plagiarism item in the sidebar loads the related submissions and displays them in the split view
3. The split view should be resizable
4. Split view shortcuts should work (top right)
5. Confirming/Denying a plagiarism should update its status indicator in the sidebar
6. Downloading the results should include a 'confirmed' column/attribute if the plagiarism was confirmed/denied (both JSON and CSV)

### Screenshots
Screenshot 2020-09-28 at 15 30 08

### CODE REVIEW

File: src/main/webapp/app/exercises/shared/plagiarism/plagiarism-split-view/plagiarism-split-view.component.html [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<div class="plagiarism-split-view">
+    <div class="plagiarism-split-pane" jhiPane>
+        <div class="plagiarism-split-pane-header">{{ comparison?.element1.studentLogin }}</div>
+        <jhi-modeling-editor *ngIf="submission1" [umlModel]="submission1?.model" [diagramType]="'ClassDiagram'" [readOnly]="true"></jhi-modeling-editor>
+    </div>
+    <div class="plagiarism-split-pane" jhiPane>
+        <div class="plagiarism-split-pane-header">{{ comparison?.element2.studentLogin }}</div>
+        <jhi-modeling-editor *ngIf="submission2" [umlModel]="submission2?.model" [diagramType]="'ClassDiagram'" [readOnly]="true"></jhi-modeling-editor>
+    </div>
+    s
```
Comment (sjagla27) [inline src/main/webapp/app/exercises/shared/plagiarism/plagiarism-split-view/plagiarism-split-view.component.html:10]: There is an extra "s" you missed.

**Review by sjagla27 — CHANGES_REQUESTED** [review]
For smaller browser windows and multiple plagiarisms the list is too long for the module and messes up the page:
grafik
Besides that and the small comment i left everything looks good.

**Comment by philippbauch** [comment]
> For smaller browser windows and multiple plagiarisms the list is too long for the module and messes up the page:
> grafik
> Besides that and the small comment i left everything looks good.

Good point! Just pushed a fix for this, the list should now be scrollable.

**Review by sjagla27 — DISMISSED** [review]
Looks good now. :)

File: package.json [hunk 2]
```diff
@@ -73,6 +73,7 @@
         "simple-statistics": "7.2.0",
         "smoothscroll-polyfill": "0.4.4",
         "sockjs-client": "1.5.0",
+        "split.js": "^1.6.2",
```
Comment (krusche) [inline package.json:76]:
```suggestion
        "split.js": "1.6.2",
```

**Comment by krusche** [comment]
The PR does not contain tests. Please add them in a follow up. As we would like to use the feature in the upcoming week, we can merge, if enough reviewers have tested and approved the PR.

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/exercises/shared/plagiarism/plagiarism-inspector/plagiarism-inspector.component.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                        comparisonResult.element2.submissionId +
+                        '/assessment?optimal=false&hideBackButton=true',
+                });
+            });
+
+            csvExporter.generateCsv(csvData);
+        }
+    }
+
+    goBack() {
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/exercises/shared/plagiarism/plagiarism-inspector/plagiarism-inspector.component.ts:114]: You do not need a method for this. Just use the routerLink directive :)

File: src/main/webapp/app/exercises/shared/plagiarism/plagiarism-inspector/plagiarism-inspector.component.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+import { ExportToCsv } from 'export-to-csv';
+
+@Component({
+    selector: 'jhi-plagiarism-inspector',
+    styleUrls: ['./plagiarism-inspector.component.scss'],
+    templateUrl: './plagiarism-inspector.component.html',
+})
+export class PlagiarismInspectorComponent implements OnDestroy, OnInit {
+    selectedComparisonIndex: number;
+    checkPlagiarismInProgress: boolean;
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/exercises/shared/plagiarism/plagiarism-inspector/plagiarism-inspector.component.ts:17]: Should this not be named something like `isPlagiarismCheckInProgress`?

File: src/main/webapp/app/exercises/shared/plagiarism/plagiarism-inspector/plagiarism-inspector.component.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    ngOnInit() {
+        this.subscription = this.route.params.subscribe((params) => {
+            this.modelingExerciseService.find(params['exerciseId']).subscribe((response: HttpResponse<ModelingExercise>) => {
+                this.modelingExercise = response.body!;
+            });
+        });
+    }
+
+    ngOnDestroy() {
+        this.subscription.unsubscribe();
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/exercises/shared/plagiarism/plagiarism-inspector/plagiarism-inspector.component.ts:35]: You do not need to unsubscribe to http subscriptions as far as i know. 
https://medium.com/@lukaonik/do-we-need-to-unsubscribe-http-client-in-angular-86d781522b99

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/exercises/shared/plagiarism/plagiarism-inspector/plagiarism-inspector.component.ts (hunk 4 above)
Comment (philippbauch) [reply src/main/webapp/app/exercises/shared/plagiarism/plagiarism-inspector/plagiarism-inspector.component.ts:17]: I don't exactly know about our preferred naming convention, but personally, I would use `isPlagiarismCheckInProgress()` as a function name (asking a question) and `plagiarismCheckInProgress` as the name of the answer to that question, containing the actual boolean value.

I could rename `checkPlagiarismInProgress` to `plagiarismCheckInProgress`, as the prefix `check` might be confusing and more suitable for a function name.

File: src/main/webapp/app/exercises/shared/plagiarism/plagiarism-header/plagiarism-header.component.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+@Component({
+    selector: 'jhi-plagiarism-header',
+    styleUrls: ['./plagiarism-header.component.scss'],
+    templateUrl: './plagiarism-header.component.html',
+})
+export class PlagiarismHeaderComponent {
+    @Input() comparisonIdx: number;
+    @Output() split = new EventEmitter<string>();
+    @Output() tag = new EventEmitter<boolean>();
```
Comment (fde312) [inline src/main/webapp/app/exercises/shared/plagiarism/plagiarism-header/plagiarism-header.component.ts:11]:
```suggestion
    @Output() onPlagiarismConfirmation = new EventEmitter<boolean>();
```

File: src/main/webapp/app/exercises/shared/plagiarism/plagiarism-header/plagiarism-header.component.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { Component, EventEmitter, Input, Output } from '@angular/core';
+
+@Component({
+    selector: 'jhi-plagiarism-header',
+    styleUrls: ['./plagiarism-header.component.scss'],
+    templateUrl: './plagiarism-header.component.html',
+})
+export class PlagiarismHeaderComponent {
+    @Input() comparisonIdx: number;
+    @Output() split = new EventEmitter<string>();
```
Comment (fde312) [inline src/main/webapp/app/exercises/shared/plagiarism/plagiarism-header/plagiarism-header.component.ts:10]:
```suggestion
    @Output() onSplitViewChange = new EventEmitter<string>();
```

**Review by fde312 — APPROVED** [review]
Works like a charm and code looks also good to me! I just made some small renaming suggestions. You can also use the routerlink directive like @stefanwaldhauser suggested. :) 

Very good job! 🚀

File: src/main/webapp/app/exercises/modeling/manage/modeling-exercise.service.ts [hunk 8]
```diff
@@ -14,6 +14,7 @@ export type EntityArrayResponseType = HttpResponse<ModelingExercise[]>;
 export type ModelingSubmissionComparisonDTO = {
     element1: ModelingSubmissionComparisonElement;
     element2: ModelingSubmissionComparisonElement;
+    confirmed?: boolean;
```
[the remark this answers is not in the document]
Comment (philippbauch) [reply src/main/webapp/app/exercises/modeling/manage/modeling-exercise.service.ts:17]: The plagiarism detection result from the backend does not include the `confirmed` attribute on plagiarisms. It is only added if either an instructor or TA has manually confirmed/denied a plagiarism. So if `confirmed` is `undefined`, it still needs manual inspection.

Maybe we could use an `enum` here:
```typescript
enum PlagiarismStatus {
    CONFIRMED = 'CONFIRMED',
    DENIED = 'DENIED',
    PENDING = 'PENDING',
}
```

File: src/main/webapp/app/exercises/shared/plagiarism/plagiarism-header/plagiarism-header.component.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (philippbauch) [reply src/main/webapp/app/exercises/shared/plagiarism/plagiarism-header/plagiarism-header.component.ts]: I updated the name of both `split` and `tag` according to @fde312's feedback 👍

File: src/main/webapp/app/exercises/modeling/manage/modeling-exercise.service.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (philippbauch) [reply src/main/webapp/app/exercises/modeling/manage/modeling-exercise.service.ts:17]: I think for now this should be fine. Can think about using an `enum` instead in a future follow-up PR 👍

**Comment by philippbauch** [comment]
Implemented the feedback from @stefanwaldhauser and @fde312 👍

File: src/main/webapp/app/exercises/modeling/manage/modeling-exercise-detail.component.ts [hunk 9]
```diff
@@ -23,7 +23,6 @@ export class ModelingExerciseDetailComponent implements OnInit, OnDestroy {
     gradingInstructions: SafeHtml;
     sampleSolution: SafeHtml;
     sampleSolutionUML: UMLModel;
-    checkPlagiarismInProgress: boolean;
```
Comment (fde312) [inline src/main/webapp/app/exercises/modeling/manage/modeling-exercise-detail.component.ts:28]: Build is failing I think because of this.

**Review by kloessst — CHANGES_REQUESTED** [review]
Feature works nicely and looks really good. Well done!

backButtonNotWorking
Just found that the back button is not working properly. I'd suggest to remove the back button unless there multiple ways to navigate to the view.

**Comment by philippbauch** [comment]
> Feature works nicely and looks really good. Well done!
> 
> backButtonNotWorking
> Just found that the back button is not working properly. I'd suggest to remove the back button unless there multiple ways to navigate to the view.

Did you pull the latest changes? I forgot to import the `RouterModule` so the `routerLink` directive didn't work properly, but this should be fixed now 👍

File: src/main/webapp/i18n/de/plagiarism.json [hunk 10]
```diff
@@ -0,0 +1,14 @@
+{
+    "artemisApp": {
+        "plagiarism": {
+            "confirm": "Bestätigen",
+            "deny": "Ablehnen",
+            "warning": "Bestätigen/Ablehnen eines Plagiats geht beim Neuladen der Seite verloren.",
```
Comment (kloessst) [inline src/main/webapp/i18n/de/plagiarism.json:6]:
```suggestion
            "warning": "Die Bestätigung/Ablehnung eines Plagiats geht beim Neuladen der Seite verloren.",
```

File: src/main/webapp/app/exercises/shared/plagiarism/plagiarism-split-view/plagiarism-split-view.component.html [hunk 11]
```diff
@@ -0,0 +1,10 @@
+<div class="plagiarism-split-view">
+    <div class="plagiarism-split-pane" jhiPane>
+        <div class="plagiarism-split-pane-header">{{ comparison?.element1.studentLogin }}</div>
+        <jhi-modeling-editor *ngIf="submission1" [umlModel]="submission1?.model" [diagramType]="'ClassDiagram'" [readOnly]="true"></jhi-modeling-editor>
```
Comment (kloessst) [inline src/main/webapp/app/exercises/shared/plagiarism/plagiarism-split-view/plagiarism-split-view.component.html:4]: ClassDiagram is hard-coded here. Do you have access to the exercise? You could use `exercise.diagramType` then.

**Review by kloessst — CHANGES_REQUESTED** [review]
Just two small remarks about the code. Overall code looks really good 💯

**Comment by kloessst** [comment]
> 
> 
> > Feature works nicely and looks really good. Well done!
> > backButtonNotWorking
> > Just found that the back button is not working properly. I'd suggest to remove the back button unless there multiple ways to navigate to the view.
> 
> Did you pull the latest changes? I forgot to import the `RouterModule` so the `routerLink` directive didn't work properly, but this should be fixed now 👍

Let me check again

File: src/main/webapp/app/exercises/shared/plagiarism/plagiarism-split-view/plagiarism-split-view.component.html (hunk 11 above)
Comment (philippbauch) [reply src/main/webapp/app/exercises/shared/plagiarism/plagiarism-split-view/plagiarism-split-view.component.html:4]: Great point, forgot about that. Added an input binding to `PlagiarismSplitViewComponent`, check the latest commit 🙂

**Review by kloessst — APPROVED** [review]

**Review by kloessst — APPROVED** [review]

**Review by fde312 — APPROVED** [review]

**Review by stefanwaldhauser — APPROVED** [review]
Well done :)

**Review by krusche — APPROVED** [review]

**Comment by anditurdiu** [comment]
Is there a limit to how many classes a diagram must have before it is detected as plagiarism? 
Submission 1: 
image
Submission 2: 
image
Result: No plagiarism found
image

**Comment by krusche** [comment]
> Is there a limit to how many classes a diagram must have before it is detected as plagiarism?

Yes, the minimumModelSize is hard coded on the server side to 5. In the future, an instructor can change this value in the client

**Pull request merged**